tests/cypress/e2e/features/masks_basics.js

```diff
@@ -167,4 +172,125 @@ context('Manipulations with masks', { scrollBehavior: false }, () => {
             cy.finishMaskDrawing();
         });
     });
+
+    describe('Empty masks actions', () => {
+        beforeEach(() => {
+            cy.removeAnnotations();
+        });
+
+        after(() => {
+            cy.removeAnnotations();
+        });
+
+        function checkEraseTools(baseTool = '.cvat-brush-tools-brush', disabled = true) {
+            cy.get(baseTool).should('have.class', 'cvat-brush-tools-active-tool');
+
+            const condition = disabled ? 'be.disabled' : 'not.be.disabled';
+            cy.get('.cvat-brush-tools-eraser').should(condition);
+            cy.get('.cvat-brush-tools-polygon-minus').should(condition);
+        }
+
+        it(
+            'Drawing a mask, fully erase it. Erase tools are blocked with empty mask. Empty shape is not created.',
+            () => {
+                const erasedMask = [{
+                    method: 'brush',
+                    coordinates: [[450, 250], [600, 400], [450, 550], [300, 400]],
+                }, {
+                    method: 'polygon-minus',
+                    coordinates: [[100, 100], [700, 100], [700, 700], [100, 700]],
+                }];
+
+                cy.startMaskDrawing();
+                cy.drawMask(erasedMask);
+
+                cy.get('.cvat-brush-tools-brush').click();
+                checkEraseTools();
+
+                cy.finishMaskDrawing();
+                cy.get('#cvat_canvas_shape_1').should('not.exist');
+            });
+
+        it('Drawing a mask, finish with erasing tool. On new mask drawing tool is reset', () => {
+            const masks = [
+                [
+                    {
+                        method: 'brush',
+                        coordinates: [[450, 250], [600, 400], [450, 550], [300, 400]],
+                    }, {
+                        method: 'polygon-minus',
+                        coordinates: [[100, 100], [400, 100], [400, 400], [100, 400]],
+                    },
+                ],
+                [
+                    {
+                        method: 'brush',
+                        coordinates: [[550, 350], [700, 500], [550, 650], [400, 500]],
+                    }, {
+                        method: 'eraser',
+                        coordinates: [[550, 350]],
+                    },
+                ],
+            ];
+            for (const [index, mask] of masks.entries()) {
+                cy.startMaskDrawing();
+                cy.drawMask(mask);
+                cy.finishMaskDrawing();
+
+                cy.get(`#cvat_canvas_shape_${index + 1}`).should('exist').and('be.visible');
+
+                cy.startMaskDrawing();
+                checkEraseTools();
+                cy.finishMaskDrawing();
+            }
+        });
+
+        it('Empty masks are deleted using remove underlying pixels feature', () => {
```

Comment:
Let's check when undo/redo in case when one object was deleted and another one was updated with remove underlying pixels.

Changes should be reverted and removed object should be restored with undo/redo.
And annotations may be saved on the server after that

Reply:
Added undo/redo checks